### DISCUSSION TITLE
Using urql with Next.js in a basic SSG setup

### DISCUSSION THREAD

**Discussion by stramel** [urql-graphql/urql#1067] [Q&A]
I was attempting to use a basic `Next.js` SSG + `urql` setup similar to how Apollo and Relay do but ran into issues with the SSR exchanges' extracted data.

Here is basically what I'm attempting to do on a page:

```js
export const getStaticProps = async () => {
  // Get client
  const client = initializeClient()

  // Perform the query
  await client.query(QUERY).toPromise()

  return {
    props: {
      // Extract the cached data and pass it as a prop
      initialUrqlState: ssrCache.extractData(), // Error: Error serializing `.initialState["2934946017"].error` returned from `getStaticProps` in "/".
      revalidate: 1,
    },
  }
}
```

then in the custom app:

```js
export default function App({ Component, pageProps: {initialUrqlState, ...pageProps }) {
  // Takes the initial state, loads it into the SSR cache, and returns the client with the state
  const urqlClient = useUrql(initialUrqlState)
  return (
    <Provider value={urqlClient}>
      <Component {...pageProps} />
    </Provider>
  )
}
```

Error Message: 

```
Error: Error serializing `.initialState["2934946017"].error` returned from `getStaticProps` in "/".
Reason: `undefined` cannot be serialized as JSON. Please use `null` or omit this value all together.
```

However, the `ssrExchange.extractData()` uses `undefined` throughout the object structure which prevents Next.js from serializing it.

If instead it were to utilize `null` or omit the `undefined` values, everything would work.

**Comment by grikomsn** [comment]
I encountered the same issue, and I have no idea how to resolve this. Probably I will create a repro repo with the same issue.

**Reply by JoviDeCroock** [reply]
What version are you using for urql/core?

**Reply by grikomsn** [reply]
In `package.json` we're using `^1.11.6`, and in `yarn.lock` it has these:

```
...
"@urql/core@>=1.14.0", "@urql/core@^1.16.0":
  version "1.16.2"
  resolved "https://registry.yarnpkg.com/@urql/core/-/core-1.16.2.tgz#e404b7f66c163245f44e7ef3c736707b7fb20234"
  integrity sha512-Be2ekFcRovpLx6EKu/w7M3jJdC32Ev3eXJgXcJD7pVDgjv/RfUbYC4tGU87VrvQDJOg+t4gHFtIR6Vr9ws7mbg==
  dependencies:
    "@graphql-typed-document-node/core" "^3.1.0"
    wonka "^4.0.14"
...
urql@^1.11.6:
  version "1.11.6"
  resolved "https://registry.yarnpkg.com/urql/-/urql-1.11.6.tgz#6b6b60093fc9ef142915938a8dbd3cc84e2dd195"
  integrity sha512-PpXzZiCKStcg52M9hJZTbcU5wIvQlYqu0vVofdaEv144JbvOhKIwzcX0+qkc3yJMyDm2isYo4D1vQQY+meRVqg==
  dependencies:
    "@urql/core" "^1.16.0"
    wonka "^4.0.14"
...
```

And the `getStaticProps` that we use is this:

```tsx
export const getStaticProps: GetStaticProps = async (ctx) => {
  const ssrCache = ssrExchange({
    isClient: false,
  });

  const client = initUrqlClient(
    {
      url: GRAPHQL_ENDPOINT,
      exchanges: [dedupExchange, cacheExchange, ssrCache, fetchExchange],
    },
    false,
  );

  await client.query(QUERY).toPromise();

  return {
    props: {
      urqlState: ssrCache.extractData(),
    },
    revalidate: 1,
  };
};
```

And on the console it shows this error:

```
SerializableError: Error serializing `.urqlState["1415180293"].error.graphQLErrors[0].path` returned from `getStaticProps` in "...".
Reason: `undefined` cannot be serialized as JSON. Please use `null` or omit this value.
    at isSerializable (/Users/griko/Workspace/.../.../node_modules/next/dist/lib/is-serializable-props.js:7:95)
...
```

And this is the resulting error on the browser:

<img width="1392" alt="image" src="https://user-images.githubusercontent.com/8220954/106044710-ab872b00-6112-11eb-863e-4766fdcc0271.png">

I tried not passing `urqlState: ssrCache.extractData()` in `getStaticProps`, and urql will just work on client-side, and if changing to another state on the same page, it shows the previous data when it's still fetching and updates to the latest data if it's finished.

Hope these help. ✌🏻

**Reply by JoviDeCroock** [reply]
This PR should've fixed that https://github.com/FormidableLabs/urql/pull/1168/files https://github.com/FormidableLabs/urql/blob/main/packages/core/src/exchanges/ssr.ts#L44 odd that, we don't make an error entry on the object if there's no error

**Reply by grikomsn** [reply]
Oh there's also this on `yarn.lock`:

```
"@urql/exchange-refocus@^0.2.1":
  version "0.2.1"
  resolved "https://registry.yarnpkg.com/@urql/exchange-refocus/-/exchange-refocus-0.2.1.tgz#0701fbe71c4009d9bcd3ae6ec7ee9ca86b70bd5d"
  integrity sha512-GDw5BbQJR6ZT3KLnrhOX6VCZWgy8MECQsR/1EOJKOWEL8luYpuUcOwYMFBdw/Vf/4ISJaA0/CVFV/cLTo6eEnA==
  dependencies:
    "@urql/core" ">=1.14.0"
    wonka "^4.0.14"
```

Does it has any effect since refetch exchange relies on `1.14.0`? Or should I inspect elsewhere?

**Reply by JoviDeCroock** [reply]
Try upgrading to latest

**Reply by grikomsn** [reply]
Did upgrade the latest, and further debugging apparently the server that we called has a GraphQL error which shows something like this:

```js
[
  {
    message: 'Unknown fragment "EntityData".',
    path: undefined,
    extensions: [Object]
  },
  ...
]
```

The `path` field which is `undefined` cannot be processed by `getStaticProps`, but after fixing the actual query, urql works now! 🎉  One solution for the error payload that I can think of is iterate the whole error object and replace `undefined` values with `null` (did try [this solution from StackOverflow](https://stackoverflow.com/a/38340374) and it worked).

Is there any way to handle `undefined` values on urql's error handling thingy?

**Reply by JoviDeCroock** [reply]
I mean, that's a next specific thing though :/ others have no issues with undefined

**Reply by grikomsn** [reply]
Fair enough. Thank you for the response! Apologies if we're asking Next.js specific issues. ✨

**Reply by JoviDeCroock** [reply]
No worries, happy to help out! Curious why they have that error rather than filtering them out
